### ISSUE TITLE
WASM: Text occasionally gets trailing garbage when building a pdf in memory 

### ISSUE THREAD

**Issue by samsieber** [ajrcarey/pdfium-render#171]
Details:
* `pdfium-render = { version = "0.8.26", default-features = false, features = ["thread_safe", "image_025", "image", "pdfium_6666"]}`
* Using the WASM build from https://github.com/paulocoutinhox/pdfium-lib/releases/tag/6684
* Reproduction repo: https://github.com/samsieber/pdfium-render-text-garbage/tree/main

I am using pdfium-render to render both normal pdfs and pdfs I generate for the purpose of generating text overlays in an image related application I'm using. When I do both, I sometimes get extra data at the end of the text I generate.

So, I might call `PdfPageTextObject::new(&document, "new test", font, font_size)` but if I later call `.text()` on said PdfPageTextObject, I could get back `"new testÿÿK"`. My theory at this point is that there's something to do with memory management going astray, but I'm not sure. 

My reproduction repository is a simplification of code I'm using in my own code. My own code targets wasm and native; I've only tried to replicate it for wasm because in my own application I've only ever seen it crop up on wasm.

Here's the core of the code that creates the text object after adding them all, checks them all. It sometimes, but not always, differs. 

```
// Add text line-by-line, handling newlines
    for line in text.lines() {
        log::info!("Creating text line: '{}'", line);
        let mut text_object = PdfPageTextObject::new(&document, line, font, font_size)?;
        text_object.set_fill_color(PdfColor::new(0, 0, 0, 255))?; // Set text color to black

        // Position the text on the page
        text_object.translate(PdfPoints::new(0.0), y_offset)?;
        y_offset -= PdfPoints::new(font_size.value * 1.4); // Adjust `y_offset` with `PdfPoints`

        // Add the text object to the page
        page.objects_mut().add_text_object(text_object)?;
    }
    for obj in page.objects().iter() {
        match &obj {
            PdfPageObject::Text(text) => {
                log::info!("Retreived text line: {}", text.text());
            }
            PdfPageObject::Path(_) => {}
            PdfPageObject::Image(_) => {}
            PdfPageObject::Shading(_) => {}
            PdfPageObject::XObjectForm(_) => {}
            PdfPageObject::Unsupported(_) => {}
        }
    }
```

The first loop is where I set the text, and the second loops checks the text; sometimes the two loops print different sets of strings, but the difference is always that there's more text than what I expect.

See the replication repository for more details; it's a modified version of the wasm example from this repository. It loads a normal pdf, renders it, and then tries to build a pdf in memory with text. If I reverse that order, the issue goes away.

**Comment by samsieber** [comment]
I have further isolated the issue. It appears to be when we copy data over. Here's a new loop I've been using:

```
    // Add text line-by-line, handling newlines
    for line in text.lines() {
        log::info!("Creating text line: '{}'", line);
        let mut text_object = PdfPageTextObject::new(&document, line, font, font_size)?;
        log::info!("Reading before attaching: '{}'", text_object.text());
        text_object.set_text(line).unwrap();
        log::info!("Reading overwritten before attaching: '{}'", text_object.text());

        text_object.set_fill_color(PdfColor::new(0, 0, 0, 255))?; // Set text color to black

        // Position the text on the page
        text_object.translate(PdfPoints::new(0.0), y_offset)?;
        y_offset -= PdfPoints::new(font_size.value * 1.4); // Adjust `y_offset` with `PdfPoints`

        // Add the text object to the page
        let mut to = page.objects_mut().add_text_object(text_object)?;
        log::info!("Reading after attaching: '{}'", to.as_text_object().unwrap().text());
        to.as_text_object_mut().unwrap().set_text(line).unwrap();
        log::info!("Reading overwritten after attaching: '{}'", to.as_text_object().unwrap().text());
    }
```

And here's the output for that:
```
Creating text line: 'new test' [pdfium_render_text_garbage.js:454:13](http://localhost:4000/pdfium_render_text_garbage.js)
Reading before attaching: '' [pdfium_render_text_garbage.js:454:13](http://localhost:4000/pdfium_render_text_garbage.js)
Reading overwritten before attaching: '' [pdfium_render_text_garbage.js:454:13](http://localhost:4000/pdfium_render_text_garbage.js)
Reading after attaching: 'new testÿÿK' [pdfium_render_text_garbage.js:454:13](http://localhost:4000/pdfium_render_text_garbage.js)
Reading overwritten after attaching: 'new test' [pdfium_render_text_garbage.js:454:13](http://localhost:4000/pdfium_render_text_garbage.js)
```

Notably, I cannot fix the text after noticing that it was added incorrectly.

**Comment by ajrcarey** [comment]
Hi @samsieber , thank you for reporting the issue, and for providing a reproduction repo. I can reproduce the problem.

It's possible to enable additional debug logging to the Javascript console when running in WASM; on line 66 of your index.html file, you can flip that `false` to `true`. Doing so immediately suggests the problem is in copy_string_to_pdfium() at https://github.com/ajrcarey/pdfium-render/blob/4eb4eb407dbf07f42f74f15e4fa9ee68523dd2a7/src/bindings/wasm.rs#L618. `FPDF_WIDESTRING`s are UTF-16 strings terminated with two `NULL`s; my immediate suspicion is that one or both of those `NULL`s are being omitted when an `FPDF_WIDESTRING` in pdfium-render's WASM memory heap is being copied across to Pdfium's WASM memory heap.

**Comment by ajrcarey** [comment]
Ok, I think I see the problem. Many characters (including standard ASCII characters) can be represented with only one byte;  in UTF-16, these are padded to two bytes with the second byte always being 0.

The current implementation of copy_string_to_pdfium() correctly scans for two consecutive NULL (i.e. 0) bytes. However, for a two-byte character that includes a second byte of 0 padding, this will result in an off-by-one error.

**Comment by ajrcarey** [comment]
`copy_string_to_pdfium()` needs to search for two consecutive NULLs _that are aligned to a two-byte boundary_. It currently doesn't, and that's the root cause here.

It also explains why the problem doesn't occur on every string; characters that don't require a second 0 padding byte won't trigger the problem.

**Comment by ajrcarey** [comment]
I have pushed a small patch that should correct the behaviour. You can test it by taking `pdfium-render` as a git dependency in your `Cargo.toml`.

**Comment by ajrcarey** [comment]
That's great. The patch will be included in crate release 0.8.27.
